pkg/controller/pxcbackup/controller.go

```diff
@@ -107,102 +129,198 @@ func (r *ReconcilePerconaXtraDBClusterBackup) Reconcile(request reconcile.Reques
-		return reconcile.Result{}, fmt.Errorf("job/setControllerReference: %v", err)
+	if err := setControllerReference(cr, job, r.scheme); err != nil {
+		return reconcile.Result{}, errors.Wrap(err, "job/setControllerReference")
 	}
 
 	err = r.client.Create(context.TODO(), job)
-	if err != nil && !errors.IsAlreadyExists(err) {
-		return reconcile.Result{}, fmt.Errorf("create backup job: %v", err)
+	if err != nil && !k8sErrors.IsAlreadyExists(err) {
+		return reconcile.Result{}, errors.Wrap(err, "create backup job")
 	} else if err == nil {
 		reqLogger.Info("Created a new backup job", "Namespace", job.Namespace, "Name", job.Name)
 	}
 
-	err = r.updateJobStatus(instance, job, destination, instance.Spec.StorageName, s3status)
+	err = r.updateJobStatus(cr, job, destination, cr.Spec.StorageName, s3status)
 
 	return rr, err
 }
 
+func (r *ReconcilePerconaXtraDBClusterBackup) tryRunS3BackupFinalizerJob(cr *api.PerconaXtraDBClusterBackup) {
+	if cr.ObjectMeta.DeletionTimestamp == nil || cr.Status.S3 == nil {
+		return
+	}
+
+	select {
+	case r.chLimit <- struct{}{}:
+		_, ok := r.bcpDeleteInProgress.LoadOrStore(cr.Name, struct{}{})
+		if ok {
+			<-r.chLimit
+			return
+		}
+
+		go r.runS3BackupFinalizer(cr)
+	default:
+		if _, ok := r.bcpDeleteInProgress.Load(cr.Name); !ok {
+			inprog := []string{}
+			r.bcpDeleteInProgress.Range(func(key, value interface{}) bool {
+				inprog = append(inprog, key.(string))
+				return true
+			})
+
+			log.Info("all workers are busy - skip backup deletion for now",
+				"backup", cr.Name, "in progress", strings.Join(inprog, ", "))
+		}
+	}
+}
+
+func (r *ReconcilePerconaXtraDBClusterBackup) runS3BackupFinalizer(cr *api.PerconaXtraDBClusterBackup) {
+	defer func() {
+		r.bcpDeleteInProgress.Delete(cr.Name)
+		log.Info("backup was removed from s3", "name", cr.Name)
+		<-r.chLimit
+	}()
+
+	s3cli, err := r.s3cli(cr)
+	if err != nil {
+		log.Error(err, "failed to create minio client for backup", "backup", cr.Name)
+	}
+
+	finalizers := []string{}
+
+	for _, f := range cr.GetFinalizers() {
+		if f == api.FinalizerDeleteS3Backup {
+			log.Info("deleting backup from s3", "name", cr.Name)
+
+			spl := strings.Split(cr.Status.Destination, "/")
+			backup := spl[len(spl)-1]
+
+			for _, bcp := range []string{backup + ".md5", backup + "sst_info", backup} {
+				err := r.removeBackup(cr.Status.S3.Bucket, bcp, s3cli)
+				if err != nil {
+					log.Error(err, "failed to delete backup", "name", cr.Name)
+					finalizers = append(finalizers, f)
+					break
+				}
+			}
+		} else {
+			finalizers = append(finalizers, f)
+		}
+	}
+
+	cr.SetFinalizers(finalizers)
+
+	err = r.client.Update(context.TODO(), cr)
+	if err != nil {
+		log.Error(err, "failed to update finalizers for backup", "backup", cr.Name)
+	}
+}
+
+func (r *ReconcilePerconaXtraDBClusterBackup) removeBackup(bucket, backup string, s3cli *minio.Client) error {
+	objs := s3cli.ListObjects(context.Background(), bucket,
```

Comment:
should we just call "delete" command, because we know the object name?

Reply:
No, we know only prefix, not whole name